ste/uploader-blobFS.go

```diff
@@ -169,22 +175,44 @@ func (u *blobFSUploader) GenerateUploadFunc(id common.ChunkID, blockIndex int32,
 	})
 }
 
+var flushThreshold int64 = (1024 * 1024 * 1024) * 20 //20GB
+//TODO: Should a flag be present for the threshold?
+var thresholdSetup bool
+
+func setupThreshold(blockSize int64) {
+	//Find the closest multiple of the block size to the flush threshold
+	m := flushThreshold % blockSize
+
+	if math.Abs(float64(m)) < math.Abs(float64((flushThreshold-m+blockSize)-flushThreshold)) {
+		flushThreshold -= m
+	} else {
+		flushThreshold = (flushThreshold - m) + blockSize
+	}
+}
+
 func (u *blobFSUploader) Epilogue() {
 	jptm := u.jptm
 
 	// flush
 	if jptm.TransferStatus() > 0 {
 		md5Hash, ok := <-u.md5Channel
+		ss := jptm.Info().SourceSize
 		if ok {
-			//Type assertions aren't my favorite thing to do but it does work when you need it to.
-			_, err := u.fileURL.FlushData(jptm.Context(), jptm.Info().SourceSize, md5Hash, *u.creationTimeHeaders)
-			if err != nil {
-				jptm.FailActiveUpload("Flushing data", err)
-				// don't return, since need cleanup below
+			// Flush incrementally to avoid timeouts on a full flush
+			for i := int64(math.Min(float64(ss), float64(flushThreshold))); i <= ss; i = int64(math.Min(float64(ss), float64(i+flushThreshold))) {
+				// Close only at the end of the file, keep all uncommitted data before then.
+				_, err := u.fileURL.FlushData(jptm.Context(), i, md5Hash, *u.creationTimeHeaders, i != ss, i == ss)
+				if err != nil {
+					jptm.FailActiveUpload("Flushing data", err)
+					break // don't return, since need cleanup below
+				}
+
+				if i == ss {
```

Comment:
Do you need this. Doesn't the for loop already take care of termination condition?

Reply:
Yes, it is actually necessary. I wrote this to exit the for loop only *after* firing one last time at the end of the file. If I dropped this, the loop would never exit, as `i` can never be more than `ss`. As such, we can have this or minor code duplication.

Reply:
Oh, I see. 
Could the for loop have therefore been written like this?  (note the "; ;" in the middle)

`for i := int64(math.Min(float64(ss), float64(u.flushThreshold))); ; i = int64(math.Min(float64(ss), float64(i+u.flushThreshold))) `

I'd suggest that such a formulation would show your intent more clearly, and be more maintainable.

Reply:
BTW, I do like the fact that this test here, at the end of the loop, makes it very clear that we are only exiting exactly when i == ss.  That makes it easier to read and understand that we really do always reach the exact end of the file.